cdap-data-fabric/src/main/java/co/cask/cdap/data/format/GrokRecordFormat.java

```diff
@@ -0,0 +1,179 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.data.format;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.format.UnexpectedFormatException;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.data.schema.UnsupportedTypeException;
+import co.cask.cdap.api.flow.flowlet.StreamEvent;
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableMap;
+import oi.thekraken.grok.api.Grok;
+import oi.thekraken.grok.api.Match;
+import oi.thekraken.grok.api.exception.GrokException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * GrokRecordFormat
+ */
+public class GrokRecordFormat extends StreamEventRecordFormat<StructuredRecord> {
+
+  public static final String PATTERN_SETTING = "pattern";
+  private static final Logger LOG = LoggerFactory.getLogger(GrokRecordFormat.class);
+  private static final String DEFAULT_PATTERN = "%{GREEDYDATA:body}";
+
+  private final Charset charset = Charsets.UTF_8;
+  private final Grok grok = new Grok();
+  private String pattern = null;
+
+  public static Map<String, String> settings(String pattern) {
+    return ImmutableMap.of(PATTERN_SETTING, pattern);
+  }
+
+  @Override
+  public StructuredRecord read(StreamEvent event) throws UnexpectedFormatException {
+    String bodyAsStr = Bytes.toString(event.getBody(), charset);
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+
+    Match gm = grok.match(bodyAsStr);
+    gm.captures();
+    Map<String, Object> x = gm.toMap();
+
+    for (Schema.Field field : schema.getFields()) {
+      String fieldName = field.getName();
+      Object value = x.get(fieldName);
+      if (value != null) {
+        builder.convertAndSet(fieldName, value.toString());
+      }
+    }
+
+    return builder.build();
+  }
+
+  @Override
+  protected Schema getDefaultSchema() {
+    // default is a String[]
+    return Schema.recordOf("streamEvent", Schema.Field.of("body", Schema.nullableOf(Schema.of(Schema.Type.STRING))));
+  }
+
+  @Override
+  protected void validateSchema(Schema desiredSchema) throws UnsupportedTypeException {
+    // a valid schema is a record of simple types. In other words, no maps, arrays, records, unions, or enums allowed.
+    // the exception is the very last field, which is allowed to be an array of simple types.
```

Comment:
As far as I know, the Grok library is simply a regex parser. So where does this restriction and exception come from? Maybe worth document it in the javadoc of this class?